modules/PowermeterBSM/lib/sunspec_base.hpp

```diff
@@ -0,0 +1,309 @@
+#include <iomanip>
+#include <map>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+using namespace std::string_literals;
+
+enum struct PointType {
+    acc32,
+    bitfield32,
+    enum16,
+    int16,
+    pad,
+    sunssf,
+    uint16,
+    uint32,
+    string,
+    blob
+};
+
+namespace InvalidPointValue {
+
+const std::uint32_t acc32{0};
+const std::uint32_t bitfield32{0x80000000};
+const std::uint16_t enum16{0xffff};
+const std::int16_t int16{static_cast<std::int16_t>(0x8000)};
+const std::uint16_t pad{0x8000};
+const std::uint16_t sunssf{0x8000};
+const std::uint16_t uint16{0xffff};
+const std::uint32_t uint32{0xffffffff};
+
+inline bool invalid_string(std::string s) {
+    return s.empty() || s.at(0) == 0; // string starts with a endmarker
+}
+
+}; // namespace InvalidPointValue
+
+// clang-format off
+const std::map<PointType,std::string> pointTypeToStringMap {
+
+    { PointType::acc32 , "acc32" },
+    { PointType::bitfield32 , "bitfield32" },
+    { PointType::enum16 , "enum16" },
+    { PointType::int16 , "int16" },
+    { PointType::pad , "pad" },
+    { PointType::sunssf , "sunssf" },
+    { PointType::uint16 , "uint16" },
+    { PointType::uint32 , "uint32" },
+    { PointType::string , "string" },
+    { PointType::blob   , "blob" },
+
+};
+// clang-format on
+
+static std::string pointTypeToString(PointType pointType) {
+
+    auto it = pointTypeToStringMap.find(pointType);
+
+    return it == pointTypeToStringMap.end() ? "unknown type" : it->second;
+}
+
+constexpr std::uint16_t point_length_in_bytes(PointType pointType) {
+
+    switch (pointType) {
+    case PointType::acc32:
+    case PointType::bitfield32:
+    case PointType::uint32:
+        return 4;
+
+    case PointType::enum16:
+    case PointType::int16:
+    case PointType::sunssf:
+    case PointType::uint16:
+    case PointType::blob: // length information must be obained from model data
+        return 2;
+
+    case PointType::string: // this is a bit hackish... a point_length_in_bytes means that the initial point list has to
+                            // provide a length value (see calc_offset for this).
+    case PointType::pad:
+        return 0;
+    }
+}
+
+struct Point {
+    const char* id;
+    PointType pointType;
+    std::uint16_t length_in_bytes = 0;
+    std::uint16_t offset = 0;
+
+    std::string to_string() const {
+        std::stringstream ss;
+        ss << "id: [" << std::setw(12) << id << " ]"
+           << " type: [" << std::setw(12) << pointTypeToString(pointType) << " ]"
+           << " length: [" << std::setw(3) << length_in_bytes << " ]"
+           << " offset: [" << std::setw(3) << offset << " ]";
+        return ss.str();
+    }
+};
+
+namespace SunspecDecoder {
+
+inline std::uint16_t uint16_at(const Types::DataVector& data, Types::DataVector::size_type offset) {
+    return be16toh((*reinterpret_cast<const std::uint16_t*>(std::addressof(data.data()[offset]))));
```

Comment:
what about (data.data() + offset) instead of std::addressof(data.data()[offset])?